### ISSUE TITLE
Unnecessary bounds check with ulong index

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#76198]
Consider:
```C#
public static string? GetValueIfInRangeOrNull(ulong ulValue, string[] names)
{
    if (ulValue < (ulong)names.Length)
        return names[(uint)ulValue];

    return null;
}
```

There doesn't appear to be a way to write this function (without `Unsafe`) that has only a single comparison/branch.  The above produces:
```assembly
; Method Program:GetValueIfInRangeOrNull(long,ref):String
G_M000_IG01:                ;; offset=0000H
       4883EC28             sub      rsp, 40

G_M000_IG02:                ;; offset=0004H
       8B4208               mov      eax, dword ptr [rdx+08H]
       448BC0               mov      r8d, eax
       4C3BC1               cmp      r8, rcx
       7610                 jbe      SHORT G_M000_IG05

G_M000_IG03:                ;; offset=000FH
       3BC8                 cmp      ecx, eax
       7313                 jae      SHORT G_M000_IG07
       8BC1                 mov      eax, ecx
       488B44C210           mov      rax, gword ptr [rdx+8*rax+10H]

G_M000_IG04:                ;; offset=001AH
       4883C428             add      rsp, 40
       C3                   ret      

G_M000_IG05:                ;; offset=001FH
       33C0                 xor      rax, rax

G_M000_IG06:                ;; offset=0021H
       4883C428             add      rsp, 40
       C3                   ret      

G_M000_IG07:                ;; offset=0026H
       E8157AC15F           call     CORINFO_HELP_RNGCHKFAIL
       CC                   int3     
; Total bytes of code: 44
```

Whereas if I instead used `Unsafe`:
```C#
public static string? GetValueIfInRangeOrNull(ulong ulValue, string[] names)
{
    if (ulValue < (ulong)names.Length)
        return Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(names), (int)ulValue);

    return null;
}
```
I get:
```assembly
; Method Program:GetValueIfInRangeOrNull(long,ref):String
G_M000_IG01:                ;; offset=0000H

G_M000_IG02:                ;; offset=0000H
       8B4208               mov      eax, dword ptr [rdx+08H]
       483BC1               cmp      rax, rcx
       7609                 jbe      SHORT G_M000_IG05

G_M000_IG03:                ;; offset=0008H
       4863C1               movsxd   rax, ecx
       488B44C210           mov      rax, gword ptr [rdx+8*rax+10H]

G_M000_IG04:                ;; offset=0010H
       C3                   ret      

G_M000_IG05:                ;; offset=0011H
       33C0                 xor      rax, rax

G_M000_IG06:                ;; offset=0013H
       C3                   ret      
; Total bytes of code: 20
```

category:cq
theme:bounds-checks
skill-level:intermediate
cost:small
impact:small

**Comment by JulieLeeMSFT** [comment]
cc @dotnet/jit-contrib.

**Comment by AndyAyersMS** [comment]
Seems like a stretch to get this fixed for 8.0, but I'll take a look.

**Comment by AndyAyersMS** [comment]
Looks like it would require a fair amount of work in VN to recognize this, unfortunately. So am going to move it out of 8.0.
